File: src/service_client_template.cpp
#include <ros/ros.h>
#include <ros_template_programs/Expression.h>
#include <ros_template_programs/Calculation.h>
#include <iostream>

namespace ros_template_programs {
    class ServiceClient {
        private:
            ros::NodeHandle nh_;
            ros::NodeHandle pnh_;
            ros::ServiceClient client_;
            ros::Timer timer_;
            int count_;
            void callbackTimer(const ros::TimerEvent&);
        public:
            ServiceClient( );
    };
}

void ros_template_programs::ServiceClient::callbackTimer(const ros::TimerEvent&) {
    ros_template_programs::Expression ex;
    ex.a = pnh_.param<double>("a", 1);
    ex.b = pnh_.param<double>("b", 1);

    if ( count_%4 == 0 ) {
        ex.calculate = "+";
    } else if ( count_%4 == 1 ) {
        ex.calculate = "-";
    } else if ( count_%4 == 2 ) {
        ex.calculate = "*";
    } else {
        ex.calculate = "/";
    }

    ros_template_programs::Calculation srv;
    srv.request.expression = ex;
    ROS_INFO("[%s] Call the calculate_two_numbers server\nExpression :%.2f %s %.2f", 
        ros::this_node::getName().c_str(), ex.a, ex.calculate.c_str(), ex.b);
    if (client_.call(srv)) {
        ROS_INFO("Result: %.2f\n", srv.response.result);
    } else {
        ROS_ERROR("Failed to call service calculate_two_numbers\n");
    }
    count_++;
    return;
}

ros_template_programs::ServiceClient::ServiceClient( ) : nh_(), pnh_("~") {
    ros::service::waitForService("calculate_two_numbers", ros::Duration(5.0));
    ROS_INFO("[%s] Connect to the service server", ros::this_node::getName().c_str());
    client_ = nh_.serviceClient<ros_template_programs::Calculation>("calculate_two_numbers");
    timer_ = nh_.createTimer(ros::Duration(1.0), &ServiceClient::callbackTimer, this);
    count_ = 0;
}

int main(int argc, char *argv[])  {
    ros::init(argc, argv, "service_client_template");
    ros_template_programs::ServiceClient srv_clt;
    ros::spin();
    return 0;
}

File: scripts/service_client_template.py
#! /usr/bin/env python
#coding:utf-8

import rospy
from ros_template_programs.msg import Expression
from ros_template_programs.srv import Calculation

class ServiceClient():
    def __init__(self):
        rospy.wait_for_service('calculate_two_numbers')
        self.client = rospy.ServiceProxy('calculate_two_numbers', Calculation)
        
    def loop_call_server(self):
        loop_rate = rospy.Rate(1)
        count = 0
        ex = Expression()
        ex.a = rospy.get_param( rospy.get_name() + '/a', 1.0)
        ex.b = rospy.get_param( rospy.get_name() + '/b', 1.0)
        while not rospy.is_shutdown():
            if count%4 == 0:
                ex.calculate = '+'
            elif count%4 == 1:
                ex.calculate = '-'
            elif count%4 == 2:
                ex.calculate = '*'
            else:
                ex.calculate = '/'
            try:
                res = self.client(ex)
                rospy.loginfo('\n [ ' + rospy.get_name()  + ' ] Result: %.2f\n', res.result )

            except rospy.ServiceException, e:
                print "Failed to call service calculate_two_numbers : %s"%e
            count += 1
            loop_rate.sleep()
        return

if __name__ == '__main__':
    try:
        rospy.init_node('service_client_template')
        
        node = ServiceClient()
        node.loop_call_server()

        rospy.spin()

    except rospy.ROSInterruptException: pass
